### DISCUSSION TITLE
Question: Removing vehicle from database

### DISCUSSION THREAD

**Discussion by Gabtoof** [teslamate-org/teslamate#2719] [Ideas]
I had been using Teslamate with 2 cars in my account. Got rid of one of the cars. Is it possible / how do I remove that car from the database?

**Comment by Dulanic** [comment]
Easiest way is to run `docker exec -it database psql -U teslamate -W teslamate `
then just do a `select * from cars;` and find your old car... look at the id and then do a `delete from cars where id = # `replacing # /w the right car id.

See https://github.com/adriankumpf/teslamate/discussions/2719#discussioncomment-3017480 as I did miss some, thanks @kodek !

**Comment by kodek** [comment]
Note that if you only delete the rows in the `cars` table, you'll run into internal errors when visiting TeslaMate's settings page. To fix this, run:

```sql
DELETE FROM car_settings WHERE id = 1234;
```

 (where 1234 is the ID removed from the `cars` table).

### To summarize:

```bash
docker exec -it my_database_container psql -U teslamate -W teslamate
```

```sql
SELECT id, vin FROM cars;
```

```
 id |        vin
----+-------------------
  1 | 5YJ3E1EA2HF123456
  2 | 5YJYGDEF7LF456789
(2 rows)
```

Assuming `5YJ3E1EA2HF123456` must be removed:

```sql
DELETE FROM cars WHERE id = 1;
DELETE FROM car_settings WHERE id = 1;
```

(Could we get a UI option? :))

**Reply by krelltunez** [reply]
Thanks for this!  I had a car in my database that I rented over a year ago, and this helped me finally remove it.

I agree, a UI option would be fantastic.

Thx!

**Reply by iKrushYou** [reply]
amazing, thank you. I deleted the car entries and started getting internal server errors and this solved it

**Comment by wasnertobias** [comment]
@adriankumpf Can we re-open this feature request? :)

**Comment by LightBringer81** [comment]
How is this still not implemented?
I mean you don't even need to own multiple vehicles to run into this problem. As soon as some of your friends or your workplace assign you to any other vehicle than yours, these will appear in the database and there is no "easy" way to remove or at least hide them.

**Comment by adriankumpf** [comment]
> How is this still not implemented? I mean you don't even need to own multiple vehicles to run into this problem. As soon as some of your friends or your workplace assign you to any other vehicle than yours, these will appear in the database and there is no "easy" way to remove or at least hide them.

Feel free to send a PR with a fix!

**Comment by CptCheesyCrust** [comment]
Sorry for the dumb question but is there a way to see what my database is called? I've tried "database" and "teslamate", but I always get the error "no such container"

I looked in the Grafana settings and under "Data Source / Teslamate" it says that the database is actually called "teslamate".

**Comment by DrMichael** [comment]
The database container is called as in docker-compose.yml, _database_ if not changed. The database to work on is called _teslamate_ (if not changed).

**Reply by CptCheesyCrust** [reply]
OK, weird. Haven't changed anything in the YML and it says _teslamate_ there too, but he just can't find it.

Or am I missing some mistake? :-/

[Screenshot](https://ibb.co/pdfPWG2)

**Reply by DrMichael** [reply]
You need to use **docker-compose**...

**Reply by vysaman** [reply]
Hi, sorry but I have the exact same question. Using docker-compose I can see that the name of the database container and the database are _database_ and _teslamate_, so no modification. But I have the same message as @CptCheesyCrust

Were you able to remove your car ?

**Reply by CptCheesyCrust** [reply]
No, i tried it with

`docker-compose exec -it teslamate psql -U teslamate -W teslamate `

but nothing happend 
![grafik](https://user-images.githubusercontent.com/62288923/206672290-da069d43-8536-457e-803b-30696bd3857e.png)

So, it looks like i'm too stupid to work with command lines and have to wait until there is option in the GUI or an Youtube Tutorial for people like me. :-D

**Comment by brianjw** [comment]
The capability to remove and indefinitely hide one of the Tesla's on my account would be very useful to me.

**Comment by Dulanic** [comment]
I finally ran into the renal use case lol. My car is in the body shop so I have a loaner from Tesla and it gets added to the account. I'd help with a PR but I don't know elixir 😭 Thankfully I found the fix earlier so I can use my own recommendation finally lol.

**Comment by supervedro** [comment]
Hello,

@CptCheesyCrust I ran into the same problem...
Here is what I did to solve it and remove all data from my previous vehicule (I run Teslamate on Oracle Private Cloud for Free)

Execute the command **_sudo docker-compose exec database psql teslamate teslamate_**
As explained above, check the id of the car to remove (in my case it was id = 1).

I wanted to fully clean the tables so I looked into each of them to see whether there were removable lined from my previous vehicule without harming the data from my second one. I ended up with the following commands to execute:

DELETE FROM cars WHERE id = 1;
DELETE FROM car_settings WHERE id = 1;
DELETE FROM charges WHERE id < 131647;
DELETE FROM charging_processes WHERE car_id = 1;
DELETE FROM drives WHERE car_id = 1;
DELETE FROM positions WHERE car_id = 1;
DELETE FROM states WHERE car_id = 1;
DELETE FROM updates WHERE car_id = 1;

WARNING! When it comes to the table charges, I did not find any car identifier so I looked for the id from my last charging session with my previous vehicule and the id of my first session with my second vehicule (i.e. 131646 & 131647)

The biggest table is positions, there were more than 5 000 000 rows to remove there.

No Stress when it comes to the execution time, it took 14 minutes to remove 40000km data of my first Tesla.

Hope this helps! So far So good on my end!

**Reply by Dulanic** [reply]
To follow up /w what you can do it, run this BEFORE you run delete from charging_processes...

```
delete from charges
where charging_process_id in (select id from charging_processes where car_id = PUT_CAR_ID_HERE *EX 1)
```
or if you already ran delete from charging_processes, you can run this:

````
delete from charges
where charging_process_id is not in (select id from charging_processes)

**Reply by CptCheesyCrust** [reply]
It worked. Finally a clear overview again. Many many thanks!

**Reply by cjchand** [reply]
Seems that `ON DELETE CASCADE` is set on these tables, as none of the tables referenced in @supervedro's comment had records for the car deleted from these commands (car.id = 1 is the car I just sold, thus wanted to remove):

```
DELETE FROM cars WHERE id = 1;
DELETE FROM car_settings WHERE id = 1;
```

This also explains why it took about 2 minutes for the queries above to complete, which seemed a bit odd at the time given each table only had 3 records.

Anyway, hope that helps someone in the future!

**Comment by natrlhy** [comment]
Thanks! I knew I had seen a discussion around this and put some notes in here, too: https://github.com/adriankumpf/teslamate/discussions/3074

**Comment by Harleyluhja** [comment]
Hi,
I've tried several times to delete a vehicle from my database, but unfortunately I can't. This thread is very confusing for me as a noob. Could someone please recap this for me? Thanks alot

**Reply by natrlhy** [reply]
Try this:

https://github.com/adriankumpf/teslamate/discussions/3074#discussioncomment-4709108

**Comment by OttoKorrect** [comment]
I hate deleting data - and storage is cheaper than ever. Maybe I want to have a look at driving data from my old Tesla in 5 years time etc etc.

An alternative is to simply hide vehicles from Grafana, this can be done by changing the dashboard variable car_id that get set from DB query. We add a list of car IDs to not include, by adding the line `WHERE id NOT IN ('1','3')`, which would exclude cars 1 & 3 from displaying.

```
SELECT name AS __text, id AS __value 
FROM cars 
WHERE id NOT IN ('1','3')
ORDER BY display_priority ASC, name ASC;
```

And Grafana nicely gives you a preview when changing this, so you don't even need to know your IDs, just change the values until you only have the car(s) you want to show.

When you save the dashboard you can't (Cannot save provisioned dashboard) - so you need to Save As, but you can then overwrite the original, and it needs to be done for each dashboard - and of course as it is custom change it will be overwritten by an update.

Instructions:
1. open a dashboard and select the cog at the top for Dashboard Settings
2. Select variables, and then select car_id
3. Edit the query - by default it is on one line, but you can format as above so it is easier to read.
4. At bottom use the **Preview of values** to confirm the cars you want are selected
5. Copy the query so you can use it in other dashboards
6. Press Update and Save As for dashboard

**Reply by maracucho1997** [reply]
This is a great solution. It takes a minute to make copies of all of the dashboards but it's worth the extra time so you can have a version of the current car(s) but still be able to view all of the historical ones too.

**Comment by MrNickIE** [comment]
Hi (NEW) team.

Please can we get a **GUI** option to 

- Hide a Car
- Delete a Car
- Merge data from one car to another (in the case you replace a car)
